gui: create PSBT with watch-only wallet

Sjors opened the pull request:
For wallets with `WALLET_FLAG_DISABLE_PRIVATE_KEYS` this makes the watch-only balance available on the send screen (including coin selection). Instead of sending a transaction it generates a PSBT.

The user can take this PSBT and process it with [HWI](https://github.com/bitcoin-core/HWI) or put it an SD card for hardware wallets that support that.

The PSBT is copied to the clipboard. This was the easiest approach; we can add a dialog later to display it, as well as an option to save to disk.

Sjors committed:
[wallet] ListCoins: include watch-only for wallets without private keys

This makes them available in GUI coin selection.

Sjors committed:
[gui] send: include watch-only

For wallets with WALLET_FLAG_DISABLE_PRIVATE_KEYS.

Sjors committed:
[wallet] add fillPSBT to interface

jb55 commented:
Concept ACK

nvk commented:
Would love to see this happen +1

instagibbs commented:
concept ACK.

I do however think a proper dialogue will be required to let the user know what's going on at a minimum rather than magic resulting in a clipboard result.

gwillen commented:
So as some of you saw, I did some related work quite awhile ago. I have a dialog with support for creating/saving/loading/signing/sending these. It's somewhat modeled after the offline signing UI in Armory. I was never _happy_ with the UI, and I think designing this to be intuitive is really hard, but I think it may already be an improvement over clipboard-only.

Let me spend a few minutes figuring out what the most recent version is that builds cleanly and works, and I will link it here so people can try it.

promag commented:
Concept ACK.

gwillen commented:
Ok, if you'd like to see my work on this, please have a look at https://github.com/gwillen/bitcoin/tree/feature-offline-v2 .

You will need to check preferences->wallet->enable offline transaction features. Then you'll get an 'Offline' menu, a new offline transactions dialog box, and a new pair of checkboxes in the 'send' dialog: "Include watchonly coins" and "create offline transaction" (which is forced on if you include watchonly). If you select the latter, the 'send' button becomes "Create unsigned" and will take you to the new dialog.

There is still some amount of jank to be ironed out, there are some XXX comments in the source and some missing error handling.

Sjors commented:
@gwillen can you make a (draft) PR with that branch? If you rebase on top of this, you can probably drop the "Enable offline transaction features" checkbox as well as the "include watchonly coins" and "create unsigned "checkboxes, in favor of only allowing the functionality with watch-only wallets.

I like how the send button text changes, so I'll steal that for this PR :-)

<img width="898" alt="Schermafbeelding 2019-09-24 om 13 49 21" src="https://user-images.githubusercontent.com/10217/65509165-219dd100-ded2-11e9-8a3e-c6c9c93dc371.png">

Once we have descriptor wallets and better HWI integration, the simple distinction between watch-only and wallets with private keys will no longer work. For example you might be part of a 2 of 3 multisig, with the keys in the wallet. Or maybe you don't have keys in the Core wallet, but your hardware wallet can sign its part via USB.

So in terms of your wizard, I would split into two wizards: one for composing & exporting, and one for loading.

<img width="910" alt="Schermafbeelding 2019-09-24 om 14 09 54" src="https://user-images.githubusercontent.com/10217/65511729-1b125800-ded8-11e9-9a77-510e0902fcef.png">

There's probably a step 0 where you attempt to (partially) sign with your own keys and with any connected devices. If that's not enough, then you go to step 1 where you can export the PSBT via clipboard or file system.

I would introduce a fresh dialog for loading, maybe co-signing and then broadcasting a transaction.

Sjors commented:
Let's continue the UX brainstom here: #16954

practicalswift commented:
I'm afraid the current version of this PR introduces a bug in `CWallet::ListCoins`.

First, note that `==` has higher precedence than the ternary conditional:

```
[cling]$ 1 == 2 ? 3 : 4
(int) 4
[cling]$ (1 == 2) ? 3 : 4
(int) 4
[cling]$ 1 == (2 ? 3 : 4)
(bool) false
```

Second, some context:

```
unsigned int IsMine(...);
bool IsWalletFlagSet(...);
int ISMINE_WATCH_ONLY = 1;
int ISMINE_SPENDABLE = 2;
```

This is being suggested in this PR:

```
if (... &&
    IsMine(it->second.tx->vout[output.n]) ==
      IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS) ?
      ISMINE_WATCH_ONLY : ISMINE_SPENDABLE) {
```

Which -- given the precedence rules -- is equivalent to:

```
if (... &&
    (IsMine(it->second.tx->vout[output.n]) ==
      IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS)) ?
      ISMINE_WATCH_ONLY : ISMINE_SPENDABLE) {
```

Which -- given that `ISMINE_WATCH_ONLY` and `ISMINE_SPENDABLE` are both non-zero integer constants (`1` and `2`) that will be evaluated in a boolean context -- is equivalent to:

```
if (... && true) {
```

Or simply:

```
if (...) {
```

:)

I think the intention was:

```
if (... &&
    IsMine(it->second.tx->vout[output.n]) ==
      (IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS) ?
      ISMINE_WATCH_ONLY : ISMINE_SPENDABLE)) {
```

Could that be the case? :)

instagibbs commented on src/wallet/wallet.cpp:
```diff
@@ -2590,7 +2590,7 @@ std::map<CTxDestination, std::vector<COutput>> CWallet::ListCoins(interfaces::Ch
             int depth = it->second.GetDepthInMainChain(locked_chain);
             if (depth >= 0 && output.n < it->second.tx->vout.size() &&
-                IsMine(it->second.tx->vout[output.n]) == ISMINE_SPENDABLE) {
+                IsMine(it->second.tx->vout[output.n]) == IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS) ? ISMINE_WATCH_ONLY : ISMINE_SPENDABLE) {
```
As @practicalswift notes you likely wanted parenths around this new clause

instagibbs commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -383,17 +386,36 @@ void SendCoinsDialog::on_sendButton_clicked()
+        CMutableTransaction mtx = CMutableTransaction{*(currentTransaction.getWtx())};
+        PartiallySignedTransaction psbtx(mtx);
+        bool complete = false;
+        const TransactionError err = model->wallet().fillPSBT(psbtx, complete, SIGHASH_ALL, false, true);
```
nit: annotate the optional bool args for clarity

instagibbs commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -383,17 +386,36 @@ void SendCoinsDialog::on_sendButton_clicked()
-    WalletModel::SendCoinsReturn sendStatus = model->sendCoins(currentTransaction);
-    // process sendStatus and on error generate message shown to user
-    processSendCoinsReturn(sendStatus);
+    bool done = false;
```
`done` isn't really helping me parse what is going on here, maybe call it `send_failure` and reverse the meaning?

instagibbs commented:
The final dialogue should have the "Yes" text replace with "Copy unsigned tx to clipboard" or something self-explanatory.

Really "Yes" should be "Send" in the average case anyways(think I might PR that change separately)

practicalswift commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -383,17 +392,36 @@ void SendCoinsDialog::on_sendButton_clicked()
+        CDataStream ssTx(SER_NETWORK, PROTOCOL_VERSION);
+        ssTx << psbtx;
+        GUIUtil::setClipboard(EncodeBase64(ssTx.str()).c_str());
+        Q_EMIT message(tr("PSBT copied"), "Copied to clipbboard", CClientUIInterface::MSG_INFORMATION);
```
Should be "clipboard" :)

Sjors committed:
[gui] watch-only wallet: copy PSBT to clipboard

Sjors commented:
Fixed the precedence bug @practicalswift found.

I added some additional copy:
<img width="391" alt="Schermafbeelding 2019-09-26 om 11 15 42" src="https://user-images.githubusercontent.com/10217/65675777-04424180-e04f-11e9-9bd2-426104c2bda8.png">

<img width="627" alt="Schermafbeelding 2019-09-26 om 11 30 25" src="https://user-images.githubusercontent.com/10217/65676974-14f3b700-e051-11e9-91d4-479afd086d30.png">

With that I don't think it's necessary to show a dialog with the PSBT text. See also #16966 for a more rigorous approach (I'm trying to keep that an easy rebase), where it will be easier to integrate @gwillen's PSBT management changes (save to disk, check progress, etc).

instagibbs commented:
"e.g., and offline Bitcoin Core wallet, or a PSBT-compatible hardware wallet"

I think that change would make me happy.

Sjors commented:
@instagibbs done (I didn't update the screenshots)

instagibbs commented:
tACK https://github.com/bitcoin/bitcoin/pull/16944/commits/da869592ecf82f8a03eb4ab2339d6bffa3dcea58

Ran both watch-only and non-watch-only modes, tested coin control in watch-only mode, tested a Core -> HWI -> Core round-trip and successfully finalized a PSBT with the desired properties.

achow101 commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -383,17 +408,35 @@ void SendCoinsDialog::on_sendButton_clicked()
+            send_failure = true;
+        }
+    }
+    if (!send_failure) {
```
This doesn't need to be outside of the `else` block since it isn't used in the no private keys case.

Sjors replied:
In the no private keys case we also clear the form. We just assume (with an `assert`) that PSBT creation can never fail, which is why `send_failure` is always `false`.

achow101 reviewed (commented):
Conecept ACK

It would be preferable if clicking "Copy to Clipboard" wouldn't close the dialog. Also, maybe show the psbt in a text box in the dialog too?

Sjors commented:
The confirmation dialog is a `QMessageBox`, which isn't very flexible. #16966 is a better place to tweak things (and I want to avoid rebase hell there).

promag commented:
What am I missing?
```
getbalances
{
  "mine": {
    "trusted": 0.00000000,
    "untrusted_pending": 0.00000000,
    "immature": 0.00000000
  },
  "watchonly": {
    "trusted": 11202.50000000,
    "untrusted_pending": 0.00000000,
    "immature": 1250.00003320
  }
}
```
And then I get "The amount exceeds your balance."

gwillen commented:
My experience was that coin selection refused to select watchonly coins when I wanted it to, even if I set watchonly to true in the coincontrol object, unless I made some more-invasive changes, linked below. I'm not sure if that's what you're seeing, @promag. 

In particular, see my changes to CWallet::ListCoins: https://github.com/gwillen/bitcoin/compare/35991b162fd142d76d83e672002846c8a40eb794...gwillen:feature-offline-v2#diff-b2bb174788c7409b671c46ccc86034bdR2362

which aren't quite the same as the ones in this PR. When I call AvailableCoins, I force on fWatchOnly in the CCoinControl object that I hand it, which the code did not previously do even if it was called for (I had to add a flag to ListCoins.) Curiously, it seems like it worked for me _without_ the spendability-related changes that this PR makes. Not sure what's up there.

Sjors commented:
@promag how did you insert the watch-only keys into that wallet? I've only tested with [HWI](https://github.com/bitcoin-core/HWI) which imports as follows:

`[{"desc": "wpkh([8...4P/0/*)#36sal9a4", "internal": false, "range": [0, 1000], "timestamp": "now", "keypool": true, "watchonly": true}, {"desc": "wpkh([80..P/1/*)#nl2rc26w", "internal": true, "range": [0, 1000], "timestamp": "now", "keypool": true, "watchonly": true}]`

What does `getaddressinfo` tell you about the coins you're trying to spend?

Does `walletcreatefundedpsbt` fail the same way?

promag commented:
@Sjors result of `getaddressinfo`:

```
{
  "address": "2N9wHFC4dSkZMu3BCspQkApJ9Pf4coHQWUF",
  "scriptPubKey": "a914b7155eadf2015c0fe4d1e17c4e29d5d58659a9d987",
  "ismine": false,
  "solvable": true,
  "desc": "sh(wpkh([528979b9]03dce9cd688fd0a9f6dac92627691eb609b8096123455676b5b57db6abaa32f666))#w8dysyn3",
  "iswatchonly": true,
  "isscript": true,
  "iswitness": false,
  "script": "witness_v0_keyhash",
  "hex": "0014528979b9bff424bdefb6e7d6cf98f692eb5a9d05",
  "pubkey": "03dce9cd688fd0a9f6dac92627691eb609b8096123455676b5b57db6abaa32f666",
  "embedded": {
    "isscript": false,
    "iswitness": true,
    "witness_version": 0,
    "witness_program": "528979b9bff424bdefb6e7d6cf98f692eb5a9d05",
    "pubkey": "03dce9cd688fd0a9f6dac92627691eb609b8096123455676b5b57db6abaa32f666",
    "address": "bcrt1q22yhnwdl7sjtmmakultvlx8kjt4448g9nrg59w",
    "scriptPubKey": "0014528979b9bff424bdefb6e7d6cf98f692eb5a9d05"
  },
  "label": "",
  "ischange": false,
  "timestamp": 1,
  "labels": [
    {
      "name": "",
      "purpose": "receive"
    }
  ]
}
```
I've used `importpubkey`.

> Does `walletcreatefundedpsbt` fail the same way?

I'll try it.

instagibbs commented:
@promag able to reproduce issue? I'm willing to help debug.

promag commented:
> @promag able to reproduce issue?

@instagibbs no, currently it's working 👀

instagibbs commented:
code review and tACK 428bdc8041e2c1479b650207110de081a91a19eb

harding commented:
@Sjors and @gwillen thanks for working on this!  I use both the GUI and a Bitcoin Core-based cold wallet, so I'm very appreciative of this type of improvement.

Nit-ish.  If I try to create an unsigned spend by clicking "Use available balance" and checking "Subtract fee from amount", my transaction shouldn't have a change output.  However, on a wallet where I haven't created a keypool, I get an error message about not having a change address.

![2019-10-09-10_09_12_440747090](https://user-images.githubusercontent.com/61096/66516633-30ad8300-ea7d-11e9-8c21-83697fa75e47.png)

I also get a second error after that error ("Transaction creation failed") which seems like error message overkill.

After importing a key to the keypool, this generated the prompt as expected.  However, I hit cancel, tried again, and it failed with the above error message.  It seems it marked the change address it allocated as used even though (1) the transaction didn't include a change address and (2) the PSBT was never sent to the clipboard (so, even if there had been a change address, it would never have been user-accessible).

After importing another single key into the keypool and this time not cancelling, it worked as expected.  I was able to take the PSBT from the clipboard, run it through my other wallet's `walletprocesspsbt`, copy it back, run `finalizepsbt`, run `sendrawtransaction`, and verify it entered the mempool of my connected regtest node.

I also tried this on a mixed wallet with some private keys and some watch-only keys, and I can confirm that it doesn't work there (as it shouldn't, per the OP).  For reference, I ran everything manually (no HWI) and only tested single-sig on regtest.

Thanks again for working on this!

Sjors commented:
The wallet seems a bit overzealous on trying to top up the keypool. I would prefer to punt this particular bug until after #16341, unless @achow101 thinks a fix won't create rebase hell for him.

promag commented:
> I get an error message about not having a change address

Yeah also got that one. @Sjors I think that, regardless of #16341, a solution would be nice.

Sjors commented:
I reproduced the issue by creating a watch-only wallet without a keypool (using HWI's `--nokeypool` feature). Added a commit to address this.

promag commented on src/wallet/wallet.cpp:
```diff
@@ -2969,7 +2969,9 @@ bool CWallet::CreateTransaction(interfaces::Chain::Lock& locked_chain, const std
-            CScript scriptChange;
+            CScript scriptChange = CScript();
+            bool failed_to_get_change_address = false;
+            std::string failed_to_get_change_address_reason = "";
```
1cb1381ca51d1967146a0a75fd96c1b4800ec378

I think you don't need this, you can write the reason in `strFailReason`.

promag commented on test/functional/wallet_keypool.py:
```diff
@@ -82,5 +82,28 @@ def run_test(self):
+        assert_raises_rpc_error(-4, "Can't generate a change-address key. No keys in the internal keypool and can't generate any keys.", w2.walletcreatefundedpsbt, inputs=[], outputs=[{addr.pop(): 0.05}], options={"subtractFeeFromOutputs": [0]})
+
+        # creating a transaction without change should still be possible
+        res = w2.walletcreatefundedpsbt(inputs=[], outputs=[{addr.pop(): 0.1}], options={"subtractFeeFromOutputs": [0]})
```
1cb1381ca51d1967146a0a75fd96c1b4800ec378

Could also test it works with `changeAddress` option?

Sjors replied:
Added

promag commented on src/wallet/wallet.cpp:
```diff
@@ -2982,21 +2984,23 @@ bool CWallet::CreateTransaction(interfaces::Chain::Lock& locked_chain, const std
+                        scriptChange = GetScriptForDestination(dest);
+                    } else {
+                        failed_to_get_change_address = true;
+                        failed_to_get_change_address_reason = "Keypool ran out, please call keypoolrefill first";
```
1cb1381ca51d1967146a0a75fd96c1b4800ec378

unrelated, I wonder why this is not translated.

Sjors replied:
Added translation

promag reviewed (commented):
Code review ACK 1cb1381ca51d1967146a0a75fd96c1b4800ec378, I'll try it out.

Sjors commented on src/wallet/wallet.cpp:
```diff
@@ -2969,7 +2969,9 @@ bool CWallet::CreateTransaction(interfaces::Chain::Lock& locked_chain, const std
             // TODO: pass in scriptChange instead of reservedest so
             // change transaction isn't always pay-to-bitcoin-address
-            CScript scriptChange;
+            CScript scriptChange = CScript();
```
So it's not undefined when passed into coin selection.

promag replied:
The empty constructor was already being called right?

Sjors replied:
Ah yes, it's a class, so it's initialized... Fixed.

achow101 reviewed (commented):
Code Review ACK 6429af6e4a41d93a632770e050effc8e3e74b023

Looked over the code and did a manual test.

> > @promag able to reproduce issue?
> 
> @instagibbs no, currently it's working eyes

I ran into this issue too, and in the past with other watch only wallet things. I think it's just due to balance caching and it should go away after closing and reopening the wallet. I'll investigate as it is not caused by this PR. edit: maybe not. I was not able to reliably reproduce afterwards.

Sjors commented on src/wallet/wallet.cpp:
```diff
@@ -3207,6 +3210,11 @@ bool CWallet::CreateTransaction(interfaces::Chain::Lock& locked_chain, const std
             }
+
+            // Give up if change keypool ran out and coin selection failed to find a solution without change:
+            if (failed_to_get_change_address && nChangePosInOut != -1) {
```
If moved this check to the start of the loop.

Sjors replied:
That didn't work. I don't like putting this deep inside hard to reason about loop & if/else branches; don't want to accidentally send change to nowhere. So I added `coin_selection_params.use_knapsack` which is set to false if `failed_to_get_change_address `, in order to skip knapsack.

Sjors replied:
This seems hopeless. There's a few scenarios where BNB isn't used:

1. When you use manual coin selection in QT 

https://github.com/bitcoin/bitcoin/blob/a22b62481aae95747830bd3c0db3227860b12d8e/src/wallet/wallet.cpp#L2674-L2682

This is the scenario @gwillen ran into: https://github.com/bitcoin/bitcoin/pull/16944#issuecomment-537321922

2. When we subtract fee from amount:

https://github.com/bitcoin/bitcoin/blob/master/src/wallet/wallet.cpp#L3011-L3013

Reverted to just letting the knapsack solver do its thing.

instagibbs replied:
I think the reduced efficiency is fine in that it's easier to read at least.

achow101 replied:
Seems like we need to enable bnb for those things.

Sjors replied:
I got part of the way there in #17219

instagibbs reviewed (commented):
Does this commit( https://github.com/bitcoin/bitcoin/pull/16944/commits/acb5891cf1b2bd478a9cd8977e2af7cad5917804 ) deserve it's own PR? Just a thought.

Sjors commented:
I extracted the keypool stuff into #17219

instagibbs commented:
Will re-review, but the diff looks good.

instagibbs commented:
re-code-review ACK https://github.com/bitcoin/bitcoin/pull/16944/commits/9808f7867d9c557ba4bc253104f47fbd7f02b668

meshcollider commented:
Code review ACK 9808f7867d9c557ba4bc253104f47fbd7f02b668

promag commented:
Sorry @Sjors, missed the notification, will review.

promag commented on src/wallet/wallet.cpp:
```diff
@@ -2590,7 +2590,9 @@ std::map<CTxDestination, std::vector<COutput>> CWallet::ListCoins(interfaces::Ch
             int depth = it->second.GetDepthInMainChain(locked_chain);
             if (depth >= 0 && output.n < it->second.tx->vout.size() &&
-                IsMine(it->second.tx->vout[output.n]) == ISMINE_SPENDABLE) {
+                IsMine(it->second.tx->vout[output.n]) ==
```
a16d2eb0d593633d6ba733d761c56c835b0e30dc

Does it make sense to do something like:
```cpp
    // some comment here
    isminetype is_mine_filter = IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS)
                              ? ISMINE_WATCH_ONLY
                              : ISMINE_SPENDABLE;

…
                IsMine(it->second.tx->vout[output.n]) == is_mine_filter) {
```

Sjors replied:
Done

promag commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -630,6 +630,9 @@ void SendCoinsDialog::useAvailableBalance(SendCoinsEntry* entry)
     }
 
+    // Include watch-only for wallets without private key
+    coin_control.fAllowWatchOnly = model->privateKeysDisabled();
```
eef2b6350684f2675989b48a2d7f79c0d8a7be6e

Does this have to be here?  Couldn't this be checked in `CWallet::AvailableCoins`?

Sjors replied:
I don't think that's a good idea. The RPC sets this based on `includeWatching`, so then we'd have to make it an `Optional`.

promag replied:
Good point.

promag commented on src/interfaces/wallet.h:
```diff
@@ -195,6 +196,13 @@ class Wallet
+    //! Fill PSBT
+    virtual TransactionError fillPSBT(PartiallySignedTransaction& psbtx,
+        bool& complete,
+        int sighash_type = 1 /* SIGHASH_ALL */,
```
2bc10122874421ba7ce379325ebc2420955b1b9b

Looks like default arguments aren't used in these interfaces, @ryanofsky coincidence or is there a reason for that?

promag replied:
Maybe for now move the defaults for wallet.cpp and just have `psbtc` and `complete` arguments for now.

promag replied:
Or even just make this `unsignedFilllPSBT(psbtx)` which `assert(complete == false)` and makes the GUI code simpler?

Sjors replied:
I think refactoring the interface can be done some other time. External signers #16546 will set `sign` to true, and the result may or may not be complete (e.g. multisig).

promag commented on src/interfaces/wallet.h:
```diff
@@ -195,6 +196,13 @@ class Wallet
         bool& in_mempool,
         int& num_blocks) = 0;
 
+    //! Fill PSBT
```
2bc10122874421ba7ce379325ebc2420955b1b9b

nit, comments have period here.

promag commented on src/interfaces/wallet.h:
```diff
@@ -14,6 +14,7 @@
 #include <functional>
 #include <map>
 #include <memory>
+#include <psbt.h>
```
2bc10122874421ba7ce379325ebc2420955b1b9b

I think you should forward declare what's necessary.

Sjors replied:
I tried, but it results in having to import stuff all over the place, and errors like this:
```
interfaces/wallet.cpp:354:20: error: no matching function for call to 'FillPSBT'
            return FillPSBT(m_wallet.get(), psbtx, complete, sighash_type, sign, bip32derivs);
                   ^~~~~~~~
./wallet/psbtwallet.h:27:28: note: candidate function not viable: cannot convert argument of incomplete type 'interfaces::PartiallySignedTransaction' to 'PartiallySignedTransaction &' for 2nd argument
NODISCARD TransactionError FillPSBT(const CWallet* pwallet,
                           ^
1 error generated.
```

promag commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -21,6 +21,7 @@
 #include <chainparams.h>
 #include <interfaces/node.h>
 #include <key_io.h>
+#include <wallet/psbtwallet.h>
```
9808f7867d9c557ba4bc253104f47fbd7f02b668

nit sort.

promag commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -319,9 +327,21 @@ void SendCoinsDialog::on_sendButton_clicked()
-    questionString.append(tr("Please, review your transaction."));
+    QString pleaseReview;
+    if (!model->privateKeysDisabled()) {
+        pleaseReview = tr("Please, review your transaction.");
```
9808f7867d9c557ba4bc253104f47fbd7f02b668

Avoid `pleaseReview` by calling `questionString.append` here and below? - like L332 above.

promag commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -373,7 +393,12 @@ void SendCoinsDialog::on_sendButton_clicked()
 
-    SendConfirmationDialog confirmationDialog(tr("Confirm send coins"), questionString, informative_text, detailed_text, SEND_CONFIRM_DELAY, this);
+    const QString confirmation = model->privateKeysDisabled() ? tr("Confirm transaction proposal") : tr("Confirm send coins");
+    QString confirmButtonText = tr("Send");
```
9808f7867d9c557ba4bc253104f47fbd7f02b668

Use ternary like above or just one `if (model->privateKeysDisabled())` for these.

promag reviewed (commented):
@Sjors Some comments for your consideration.

promag commented on src/interfaces/wallet.cpp:
```diff
@@ -344,6 +345,14 @@ class WalletImpl : public Wallet
+        bool sign = true,
+        bool bip32derivs = false) override
+    {
+            return FillPSBT(m_wallet.get(), psbtx, complete, sighash_type, sign, bip32derivs);
```
d474156f242dc900c221924765210ac4f7e97f7a

Wrong indentation.

Sjors replied:
Oops, fixed.

Sjors replied:
Fixed

Sjors replied:
Fixed

Sjors commented:
Mmm, this in an odd Travis error:
```
2019-11-06T17:49:26.056000Z TestFramework (INFO): sendrawtransaction with missing input
198162019-11-06T17:49:42.869000Z TestFramework (ERROR): JSONRPC error
19817Traceback (most recent call last):
19818  File "/home/travis/build/bitcoin/bitcoin/build/bitcoin-i686-pc-linux-gnu/test/functional/test_framework/test_framework.py", line 111, in main
19819    self.run_test()
19820  File "/home/travis/build/bitcoin/bitcoin/build/bitcoin-i686-pc-linux-gnu/test/functional/rpc_rawtransaction.py", line 354, in run_test
19821    self.nodes[2].sendrawtransaction(rawTxComb)
19822  File "/home/travis/build/bitcoin/bitcoin/build/bitcoin-i686-pc-linux-gnu/test/functional/test_framework/coverage.py", line 47, in __call__
19823    return_val = self.auth_service_proxy_instance.__call__(*args, **kwargs)
19824  File "/home/travis/build/bitcoin/bitcoin/build/bitcoin-i686-pc-linux-gnu/test/functional/test_framework/authproxy.py", line 141, in __call__
19825    raise JSONRPCException(response['error'], status)
19826test_framework.authproxy.JSONRPCException: non-mandatory-script-verify-flag (Signature must be zero for failed CHECK(MULTI)SIG operation) (-26)
```

luke-jr commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -186,6 +187,13 @@ void SendCoinsDialog::setModel(WalletModel *_model)
 
+        if (model->privateKeysDisabled()) {
+            ui->sendButton->setText(tr("Create Unsigned"));
+            ui->sendButton->setToolTip(tr("Creates a Partially Signed Bitcoin Transaction (PSBT) for use with e.g. an offline Bitcoin Core wallet, or a PSBT-compatible hardware wallet."));
```
"Bitcoin Core" should be replaced with %1 and `PACKAGE_NAME`

luke-jr commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -186,6 +187,13 @@ void SendCoinsDialog::setModel(WalletModel *_model)
 
+        if (model->privateKeysDisabled()) {
+            ui->sendButton->setText(tr("Create Unsigned"));
+            ui->sendButton->setToolTip(tr("Creates a Partially Signed Bitcoin Transaction (PSBT) for use with e.g. an offline Bitcoin Core wallet, or a PSBT-compatible hardware wallet."));
```
Do we want to imply the Bitcoin Core GUI can sign these yet?

Sjors replied:
In reference to "offline Bitcoin Core wallet"? It doesn't say GUI :-)

luke-jr replied:
Hence "imply".

Sjors replied:
I only expect people to use this feature (watch only wallets in general) based on some tutorial, so I don't think this needs to be perfect. Hopefully we can add signing PSBT from the GUI soon too.

luke-jr commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -186,6 +187,13 @@ void SendCoinsDialog::setModel(WalletModel *_model)
+            ui->sendButton->setText(tr("Create Unsigned"));
+            ui->sendButton->setToolTip(tr("Creates a Partially Signed Bitcoin Transaction (PSBT) for use with e.g. an offline Bitcoin Core wallet, or a PSBT-compatible hardware wallet."));
+        } else {
+            ui->sendButton->setText(tr("Send"));
```
Do we need this? Each wallet gets its own `SendCoinsDialog`, and can't change `privateKeysDisabled` AFAIK...

Sjors replied:
Not sure what you mean here. We set these strings in `setModel(WalletModel *_model)`. That's the first time we know `privateKeysDisabled`. The constructor doesn't know.

luke-jr replied:
I mean there is never a case where we need to change the text *back* to "Send" (which is set in the .ui)

Sjors replied:
Ah I see. Ok, I dropped the else branch (keeping the original in the `.ui` file).

luke-jr commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -186,6 +187,13 @@ void SendCoinsDialog::setModel(WalletModel *_model)
+            ui->sendButton->setText(tr("Create Unsigned"));
+            ui->sendButton->setToolTip(tr("Creates a Partially Signed Bitcoin Transaction (PSBT) for use with e.g. an offline Bitcoin Core wallet, or a PSBT-compatible hardware wallet."));
+        } else {
+            ui->sendButton->setText(tr("Send"));
```
[If we do need it,] remove duplicate from .ui file

Sjors replied:
Ok, I removed the tooltip and button text from the UI file. I also added the `s` keyboard shortcut back, though I can't test that on macOS.

luke-jr commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -186,6 +187,13 @@ void SendCoinsDialog::setModel(WalletModel *_model)
+            ui->sendButton->setText(tr("Create Unsigned"));
+            ui->sendButton->setToolTip(tr("Creates a Partially Signed Bitcoin Transaction (PSBT) for use with e.g. an offline Bitcoin Core wallet, or a PSBT-compatible hardware wallet."));
+        } else {
+            ui->sendButton->setText(tr("Send"));
```
[If we do need it,] should be "S&end" for the shortcut key.

luke-jr commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -305,9 +313,19 @@ void SendCoinsDialog::on_sendButton_clicked()
 
-    QString questionString = tr("Are you sure you want to send?");
+    QString questionString;
+    if (!model->privateKeysDisabled()) {
```
Please handle the true case first.

Sjors replied:
Done

luke-jr commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -305,9 +313,19 @@ void SendCoinsDialog::on_sendButton_clicked()
+    if (!model->privateKeysDisabled()) {
+        questionString.append(tr("Please, review your transaction."));
+    } else {
+        questionString.append(tr("Please, review your transaction proposal. This will produce a Partially Signed Bitcoin Transaction (PSBT) which you can copy and then sign with e.g. an offline Bitcoin Core wallet, or a PSBT-compatible hardware wallet."));
```
Not sure we need to change this message.

Sjors replied:
This is the only place where it's not in a toolbox.

luke-jr replied:
???

Sjors replied:
I mean tool tip.

luke-jr commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -908,11 +950,11 @@ void SendConfirmationDialog::updateYesButton()
     {
         yesButton->setEnabled(false);
-        yesButton->setText(tr("Send") + " (" + QString::number(secDelay) + ")");
+        yesButton->setText(confirmButtonText + " (" + QString::number(secDelay) + ")");
```
Suggest using 4e10bdd433f5d1ba9069f4942cb89e69dd83d816 for this

Sjors replied:
I'll rebase if that's merged first, but that's not a trivial enough commit.

luke-jr commented on src/wallet/wallet.cpp:
```diff
@@ -2160,7 +2160,7 @@ std::map<CTxDestination, std::vector<COutput>> CWallet::ListCoins(interfaces::Ch
     for (const COutput& coin : availableCoins) {
         CTxDestination address;
-        if (coin.fSpendable &&
+        if ((coin.fSpendable || (IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS) && coin.fSolvable)) &&
```
I'd make a `const bool include_watch_only` to check the wallet flag once outside the loops...

luke-jr replied:
Something like this above:

```C++
    // Include watch-only for wallets without private keys
    const bool include_watch_only = IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS);
    const isminetype is_mine_filter = include_watch_only ? ISMINE_WATCH_ONLY : ISMINE_SPENDABLE;
```

luke-jr replied:
You didn't resolve this...

luke-jr commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -186,6 +187,14 @@ void SendCoinsDialog::setModel(WalletModel *_model)
+            ui->sendButton->setText(tr("&Create Unsigned"));
+            ui->sendButton->setToolTip(tr("Creates a Partially Signed Bitcoin Transaction (PSBT) for use with e.g. an offline %1 wallet, or a PSBT-compatible hardware wallet.").arg(PACKAGE_NAME));
+        } else {
+            ui->sendButton->setText(tr("&Send"));
```
Should be `"S&end"`

luke-jr commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -186,6 +187,11 @@ void SendCoinsDialog::setModel(WalletModel *_model)
         ui->optInRBF->setCheckState(Qt::Checked);
 
+        if (model->privateKeysDisabled()) {
+            ui->sendButton->setText(tr("C&reate Unsigned"));
```
Perhaps use `"Cr&eate Unsigned"` so the "E" shortcut matches with `"S&end"`

Sjors replied:
Done

instagibbs commented:
test and code review ACK https://github.com/bitcoin/bitcoin/pull/16944/commits/c6dd565c8820aa8a98b190621e10c6e2821a9ecc

jonasschnelli commented:
Concept ACK

meshcollider reviewed (commented):
re-ACK c6dd565c8820aa8a98b190621e10c6e2821a9ecc

andronoob commented:
Is it possible to create multi-part/fountain code QR code?

I don't want to see compromises like Electrum had done before reappear on Bitcoin Core...

promag commented:
I've used the "Use available balance" which filled the amount, but below balance is still zero. 

<img width="1034" alt="Screenshot 2019-11-24 at 21 27 50" src="https://user-images.githubusercontent.com/3534524/69501927-9193fc00-0f01-11ea-95c0-f84ab12675a6.png">

Something that can be fixed in a follow up - the button could change and/or the balance info on the bottom too.

Tested ACK c6dd565c8820aa8a98b190621e10c6e2821a9ecc.

Sjors commented:
@promag the balance seems unrelated, though obviously a bug

Sjors commented:
@promag fixed in #17587
